src/Neo/SmartContract/KeyBuilder.cs

```diff
@@ -21,18 +22,20 @@ namespace Neo.SmartContract
     /// </summary>
     public class KeyBuilder
     {
-        private readonly MemoryStream stream = new();
+        private readonly MemoryStream stream;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="KeyBuilder"/> class.
         /// </summary>
         /// <param name="id">The id of the contract.</param>
         /// <param name="prefix">The prefix of the key.</param>
-        public KeyBuilder(int id, byte prefix)
+        /// <param name="keySizeHint">The hint of the storage key size(not including the id and prefix).</param>
+        public KeyBuilder(int id, byte prefix, int keySizeHint = ApplicationEngine.MaxStorageKeySize)
         {
-            var data = new byte[sizeof(int)];
+            Span<byte> data = stackalloc byte[sizeof(int)];
             BinaryPrimitives.WriteInt32LittleEndian(data, id);
 
+            stream = new(sizeof(int) + sizeof(byte) + keySizeHint);
```

Comment:
`byte` shouldn't be necessary, in general `MaxStorageKeySize` limits the whole key with prefix included. That said, `KeyBuilder` is somewhat special in that only native contracts use it. And native contracts can do things normal contracts can't (like https://github.com/neo-project/neo/issues/3596). Still, current native contracts never go beyond this limit.

Reply:
> `byte` shouldn't be necessary, in general `MaxStorageKeySize` limits the whole key with prefix included. That said, `KeyBuilder` is somewhat special in that only native contracts use it. And native contracts can do things normal contracts can't (like #3596). Still, current native contracts never go beyond this limit.

fixed.